server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -1047,13 +1049,25 @@ private static <T extends ClusterStateTaskListener> ClusterState innerExecuteTas
             // to avoid leaking headers in production that were missed by tests
 
             try {
-                return executor.execute(
+                final var updatedState = executor.execute(
                     new ClusterStateTaskExecutor.BatchExecutionContext<>(
                         previousClusterState,
                         executionResults,
                         threadContext::newStoredContext
                     )
                 );
+                if (updatedState.version() != previousClusterState.version()
+                    || updatedState.routingTable().version() != previousClusterState.routingTable().version()
+                    || updatedState.metadata().version() != previousClusterState.metadata().version()) {
```

Comment:
I wonder if this should always be checked or assertion would be enough?

Reply:
The bug fixed here involved a race condition, I worry it'd never be caught in tests. Or put differently I expect it should have been caught already. And the consequences are rather severe, so it seems best to protect against it in production too.